Bash completion

RonaldBarnes opened the pull request:
* Resolves: #35450 <!-- related github issue -->

## Summary
Add bash completion for `occ`

## TODO

- [x] Support more OS's `bash_completion.d` directory locations
- [x] Add any missed web server user names (i.e. nginx)

## Checklist

- [X] Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting) **N/A for bash scripts**?
- [X] [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required (**requires confirmation of pending acceptance**, then this PR's final state will be documented and a PR to nextcloud/documentation will be made)
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes) (Is this a good idea? Wouldn't hurt and easy to integrate.)

RonaldBarnes committed:
Auto TAB completion for bash using alias for occ.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Updated copyright per Contribution guidelines.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes commented:
An example of the script in action, showing user interaction required before any action taken:

```
# ./bash-add-alias.sh
Web server user name: "www-data".
Alias for occ command not found for user "root".
Run "alias occ='sudo --user www-data php /var/www/nextcloud/occ'" (y/N)? N
Add alias to /root/.bash_aliases? (y/N) N
Add alias to /home/ron/.bash_aliases? (y/N) N
Run bash completion script complete.occ?  (Y/n) N
Add complete.occ to /etc/bash_completion.d? (y/N) N
DONE.
```

RonaldBarnes commented:
Works from any folder, even when `occ` is not in the current directory:

```
/tmp/nextcloud-occ-bash-completion# occ
activity:        encryption:      memories:        text:
app:             federation:      notification:    theming:
background:      files:           onlyoffice:      trashbin:
background-job:  group:           preview:         twofactorauth:
broadcast:       help             recognize:       update:
check            integrity:       security:        upgrade
circles:         l10n:            serverinfo:      user:
config:          list             sharing:         versions:
dav:             log:             status           workflows:
db:              maintenance:     tag:
deck:            maps:            talk:
root@(dell):/tmp/nextcloud-occ-bash-completion# occ
```

Reduce number of choices by typing `ta`:
```
/tmp/nextcloud-occ-bash-completion# occ ta
tag:   talk:
root@(dell):/tmp/nextcloud-occ-bash-completion# occ ta
```

Increase choices by typing `l`, which completes to `talk:`, then provides all `talk:` options:
```
/tmp/nextcloud-occ-bash-completion# occ talk:
talk:active-calls         talk:room:delete          talk:stun:add
talk:command:add          talk:room:demote          talk:stun:delete
talk:command:add-samples  talk:room:promote         talk:stun:list
talk:command:delete       talk:room:remove          talk:turn:add
talk:command:list         talk:room:update          talk:turn:delete
talk:command:update       talk:signaling:add        talk:turn:list
talk:room:add             talk:signaling:delete     talk:user:remove
talk:room:create          talk:signaling:list
root@(dell):/tmp/nextcloud-occ-bash-completion# occ talk:
```

And so on:
```
/tmp/nextcloud-occ-bash-completion# occ talk:command:
talk:command:add          talk:command:delete       talk:command:update
talk:command:add-samples  talk:command:list
root@(dell):/tmp/nextcloud-occ-bash-completion# occ talk:command:list
```

RonaldBarnes commented:
Works with options beginning with `-` too, if user types in `occ -`[TAB]:
```
/tmp/nextcloud-occ-bash-completion# occ -
--ansi            --no-ansi         --quiet           --version
-h                --no-interaction  -v                -vv
--help            --no-warnings     -V                -vvv
-n                -q                --verbose
root@(dell):/tmp/nextcloud-occ-bash-completion# occ -
```

artonge commented on bash-add-alias.sh:
```diff
@@ -0,0 +1,330 @@
+				echo "N"
+			fi
+		fi
+#	elif [[ -d /usr/share/bash-completion ]] ; then
+#		if [[ -f /usr/share/bash-completion/completions/complete.occ ]] ; then
+#			script_found=0
+#			echo -n "Found existing complete.occ in "
+#			echo "/usr/share//bash-completion/completions"
+#		fi
+#		fi
```
To remove ?

artonge commented on complete.occ:
```diff
@@ -0,0 +1,148 @@
+		COMPREPLY=($(printf '%s' "${candidates[@]}"))
+	fi
+
+	## TAB=9, ALT+?=63,...
+	## echo "COMP_TYPE: \"${COMP_TYPE}\""
```
To remove ?

artonge reviewed (changes requested):
Comments on current behavior:

```shell
# 1
# Comment: Is having a triling ':' expected ? Not sure if we should remove it or not.
# Having it is weird, but it suggests that occ is expecting something after the ':'.
# How are other software handling that ?
$ occ<tab>
activity:        dav:             integrity:       preview:         text:            versions:
app:             db:              l10n:            security:        theming:         workflows:
background:      encryption:      list             serverinfo:      trashbin:        
background-job:  federation:      log:             sharing:         twofactorauth:   
broadcast:       files:           maintenance:     status           update:          
check            group:           migrations:      support:         upgrade          
config:          help             notification:    tag:             user:  

# 2
$ occ files:<tab>
files:cleanup                    files:repair-tree                files:scan-app-data
files:recommendations:recommend  files:scan                       files:transfer-ownership

# 3
# Comment: I don't think have the global list of command makes sense here.
# I would maybe expect completion on the options.
$ occ files:scan <tab>
activity:        dav:             integrity:       preview:         text:            versions:
app:             db:              l10n:            security:        theming:         workflows:
background:      encryption:      list             serverinfo:      trashbin:        
background-job:  federation:      log:             sharing:         twofactorauth:   
broadcast:       files:           maintenance:     status           update:          
check            group:           migrations:      support:         upgrade          
config:          help             notification:    tag:             user:

artonge commented on complete.occ:
```diff
@@ -0,0 +1,148 @@
+	while read -r value; do
+		opts_list+=("${value}")
+		## Read output of "occ" and get word-chars following "-" or "--":
+		done <<< $(occ |
```
Do we have a reference to the command being completed here ?
I think it would be better to use the `occ` from the command instead of using the global aliased one.

RonaldBarnes replied:
Hi artonge,

Not sure I understand.

It's running `occ`, the alias, as `occ` the script is not executable and would require the full `sudo -u ... php .../occ`.

If the alias for `occ` isn't working, then this code will never get invoked.

artonge replied:
Ah ok, so we can't autocomplete when using `php ./occ` ?

RonaldBarnes replied:
No, that would require completion on `php`, I don't think it can be done on multiple words, at least not without breaking every other tab completion in bash.

Also:
```
Console has to be executed with the user that owns the file config/config.php
```

RonaldBarnes replied:
Interesting: I just tried `./occ [tab][tab]` and ... completion worked.

However:
* completion did not work on completing the `./occ` command itself, unlike the alias `occ` where `oc[tab]` does complete
* error thrown, since `./occ` is not executable:
```
./occ files:scan User1
-bash: ./occ: Permission denied
```

RonaldBarnes commented on complete.occ:
```diff
@@ -0,0 +1,148 @@
+	## Just manually add -vv and -vvv, and place in order (after -v):
+	local len
+	((len=${#opts_list[@]}-1))
+	opts_list[len]="-vv"
```
Inserting the `-vv` like that replaces the existing `--verbose`.

The following 2 lines add `-vvv` then re-add `--verbose`.

This ensures the options are listed in order:
* `-v`
* `-V`
* `-vv`
* `-vvv`
* `--verbose`

Instead of:
* `-v`
* `-V`
* `--verbose`
* `-vv`
* `-vvv`

The `-V` is slightly out of sequence, maybe, depending on collation and aesthetic opinion.

artonge replied:
Ok, so the code expects that `--verbose` is the last entry, which is probably always the case.

RonaldBarnes replied:
(I'm hoping this reply is attached to the correct comment, @artonge.)

> ```
> # 3
> # Comment: I don't think have the global list of command makes sense here.
> # I would maybe expect completion on the options.
> $ occ files:scan <tab>```

I do like this idea and have begun implementing it, but finding some issues.

Just one of them: the output of `--help` lists app-specific options, plus all of `occ`'s options.

For example, the output of `occ files:scan --help` shows 6 app-specific options, plus `occ`'s.

Present all of them, or just the ones up to `-h,  --help`?  Include the `help` ones?  The `--quiet` ones?

Then, some of the `occ` options cancel out the rest of the command entirely:
```
occ files:scan --shallow -V userX
Nextcloud 24.0.7
```

No files are scanned.


Another issue is that `--output=json` is listed as the first option, but throws an uncaught error on `occ talk:command:list --output=json`, and does nothing on several other commands:

```
occ files:scan --shallow --output=json userX
Starting scan for user 1 out of 1 (userX)
+---------+-------+--------------+
| Folders | Files | Elapsed time |
+---------+-------+--------------+
| 2       | 0     | 00:00:02     |
+---------+-------+--------------+
```

This isn't an issue for the bash completion project, but might be an issue for a whole lot of apps that claim to support json output.

RonaldBarnes replied:
> Why can't it be...

Looking into this further, the `complete` built-in handles de-duplication and sortation of completion options, so I've revised this code with commit a86af42e22447154398716af35779702d4716c06 to resolve this similar to your proposal.

RonaldBarnes commented on bash-add-alias.sh:
```diff
@@ -0,0 +1,330 @@
+				grep occ ${home_dir}/.bash_aliases
+			fi
+		else
+	##	if [[ ${answer} != "" ]] ; then
```
Good catch, will remove.

Thank you.

juliushaertl commented:
Isn't this something that symfony already provides us with as documented in https://github.com/nextcloud/documentation/blob/eae31ad02fa83e218d693f7a79eded3d127e7406/admin_manual/configuration_server/occ_command.rst#enabling-autocompletion ?

RonaldBarnes commented on bash-add-alias.sh:
```diff
@@ -0,0 +1,330 @@
+## Cleanup all variables on exit:
+## trap 'cleanup_vars' SIGINT SIGKILL SIGTERM
+## Cleanup ALL variables on exit (including cleanup_vars() itself):
+trap 'cleanup_vars ALL' RETURN EXIT QUIT SIGINT SIGKILL SIGTERM
```
Yes, if they aren't cleaned up, the vars pollute the settings, but more importantly, the RETURN trap will run when other scripts finish, thereby re-running `cleanup_vars`.

That really makes a mess of things, giving unexpected results in places unrelated to the current scripts.

RonaldBarnes committed:
Resolved requested changes to PR.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes commented on bash-add-alias.sh:
```diff
@@ -0,0 +1,330 @@
+	if [[ ! -f ${occPath} ]] ; then
+		getOccPath
+	fi
+	## echo "occPath: \"${occPath}\""
```
Not sure if I should comment when resolving?

This issue (plus several others) have been resolved.

RonaldBarnes committed:
Resolved one more comment deletion, per PR change request(s).

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes commented:
It's odd that there are a couple comments here that I can't seem to reply to.

Is this normal; am I missing something?

RonaldBarnes commented:
@juliushaertl 

> Isn't this something that symfony already provides us with as documented in https://github.com/nextcloud/documentation/blob/eae31ad02fa83e218d693f7a79eded3d127e7406/admin_manual/configuration_server/occ_command.rst#enabling-autocompletion ?

According to https://github.com/stecman/symfony-console-completion/issues/71, it doesn't seem work for everyone.

The author stated:
> to be honest this seems like a permissions / design issue with Nextcloud, not the completion system. A few suggestions for working around this:
> 
>    Configure file/dir permissions so that www-data and the owning user can both write to those directories:

Really don't like the idea of changing file/dir permissions within the NC folder to get bash comlpletion working!  I imagine a lot of `chmod 0777` commands being run.


Also - the method at the link you provided doesn't work for me - I copied and pasted the "BASH (any version)" instructions:
```
/var/www/nextcloud# eval $(/var/www/html/nextcloud/occ _completion --generate-hook)
-bash: /var/www/html/nextcloud/occ: No such file or directory
```
Instructions point to different NC directory, adjusting for local instance: 
```
/var/www/nextcloud# ls -l occ
-rw-r--r-- 1 www-data www-data 283 2022-11-08 03:04 occ
root@(dell):/var/www/nextcloud# eval $(/var/www/nextcloud/occ _completion --generate-hook)
-bash: /var/www/nextcloud/occ: Permission denied
/var/www/nextcloud#
```
User `root` has permission denied.  Trying again as user `www-data`:
```
/var/www/nextcloud# sudo -u www-data eval $(/var/www/nextcloud/occ _completion --generate-hook)
-bash: /var/www/nextcloud/occ: Permission denied
sudo: eval: command not found
/var/www/nextcloud#
```
Re-created my own `occ` alias, then:
```
/var/www/nextcloud# occ status
- installed: true
- version: 24.0.7.1
```





That completion method is not very discoverable - I know multiple NC admins (not full-time admins, but regulars) that had no idea this Symfony method existed.

This version is discoverable and functions like every other bash completion.

RonaldBarnes committed:
Changed getHttpdUser to searchHttpdUser through-out, per PR change requests.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes commented:
Hi again @artonge 

I couldn't reply directly to your comment, so... "Quote Reply" is here:


> Comments on current behavior:
> 
> ```shell
> # 1
> # Comment: Is having a trailing ':' expected ?
Yes, it's an indicator that further subcommands are expected.

> Not sure if we should remove it or not.

I strongly recommend keeping it; it's quite informative.

> Having it is weird, but it suggests that occ is expecting something after the ':'.

Precisely.

> How are other software handling that ?

I am not aware of any other program that expects colon-delimited input like `occ`.

>```
> #3
> # Comment: I don't think have the global list of command makes sense here.
> # I would maybe expect completion on the options.

I like the idea, but I see some problems with implementation.

- Running the command to capture and parse the output looking for options would run the command, and capture the output for processing.
If no further options expected by the command, then the command has run and output is hidden from user.
- Some commands have quite long run-times, so things will appear to have hanged from user's perspective
- Idempotence might be imperilled - critical output could be captured and re-running command might return different results.

I do think I could prevent displaying the global list of commands if there is already a word with a colon in it.

Also, if there were a list of commands that expect files as final arguments, I could make it so that those commands, followed by [tab], present a list of files and folders from the current directory.

I'm unaware of any such commands that operate on files though.


> ```
> $ occ files:scan <tab>
> activity:        dav:             integrity:       preview:         text:            versions:
> app:             db:              l10n:            security:        theming:         workflows:
> background:      encryption:      list             serverinfo:      trashbin:        
> background-job:  federation:      log:             sharing:         twofactorauth:   
> broadcast:       files:           maintenance:     status           update:          
> check            group:           migrations:      support:         upgrade          
> config:          help             notification:    tag:             user:      
> ```

RonaldBarnes committed:
Revised per pull request comment https://github.com/nextcloud/server/pull/35451#discussion_r1033380070 as the 'complete' built-in handles de-duplication and sortation.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Reset COMP_WORDBREAKS to original value when finished.
Add space after completed word unless it contains a colon.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Add quotes around COMP_WORDBREAKS, else unexpected behaviour.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Fixed adding '-vv' option to array: remnant overlooked in commit a86af42e22447154398716af35779702d4716c06

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Added more web server user names.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Add user_id to completions.
Add apps list to completions.
Do not offer a switch as completion if it has already been chosen.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Web server lighthttp user name is fixed to lighttpd.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Added Usage: and remove unused debugging.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Added user_id, app, lang, and file system completions.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes commented:
Completions are now customized to the expectations of the command being run, i.e. a `user_id`:

```
occ files:scan [tab][tab]
admin  A   J1   J2


occ files:scan J[tab][tab]
J1  J2
```

A user that's been chosen / selected will be filtered from further options:
```
occ files:scan J1 [tab][tab]
admin  A   J2
```

If `-p` or `--path` is the previous word, expect a file system entry:
```
root@(dell):/var/www/nextcloud# occ files:scan --path
3rdparty/          core/              occ                status.php
apps/              cron.php           ocm-provider/      test.sh
...
```
Note: in the case of `files:scan --path`, it **must** be run from `../nextcloud/data` for file completion to give a valid option (a user's directory), **or** the argument to `--path` **must** be prefixed by `/`: this is not enforced by the completion script.

RonaldBarnes commented:
Completions are now customized to the expectations of the command being run, i.e. an `app`:

```
occ talk:command:list
accessibility              memories
activity                   nextcloud_announcements
admin_audit                notes
...
```

Lengthy list of all apps has pagination handled as any other bash completion (i.e. `ls -l /usr/bin/[tab][tab]`).

```
occ talk:command:list user_[tab][tab]
user_ldap    user_status
```

RonaldBarnes commented:
Completions are now customized to the expectations of the command being run, i.e. an `app` followed by a `lang` (language):

```
occ l[tab][tab]
l10n:  list   log:


occ l1[tab]
occ l18n:[tab]
occ l10n:createjs[tab][tab]
accessibility              memories
activity                   nextcloud_announcements
admin_audit                notes
apporder                   notifications
breezedark                 oauth2
...
```
Having chosen an app, now `l18n` expects a language **file**:
```
occ l10n:createjs user_ldap [tab][tab]
3rdparty/          core/              occ                status.php
apps/              cron.php           ocm-provider/      test.sh
AUTHORS            data/              ocs/               themes/
...
```

RonaldBarnes committed:
Changed all functions to have '_occ_' prefix.
Tidied up 'compopt' settings.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

MichaIng commented on bash-add-alias.sh:
```diff
@@ -0,0 +1,309 @@
+## Leave no trace after exit (except alias(es)):
+function cleanup_vars()
+	{
+	unset value
+	unset httpdUser
+	unset user_name
+	unset home_dir
+	unset aliasExists
+	unset phpFound
+	unset answer
+	unset aliasString
+	unset addAlias
+	unset aliasExists
+
+	unset searchHttpdUser
+	unset occOwner
+	unset occPath
+	unset getOccPath
+	unset script_found
+
+	unset green
+	unset yellow
+	unset red
+	unset default_colour
```
I suggest to use `unset -v` resp. `unset -f `, to unset variables resp. functions explicitly and rule out that a variable with the same name is unset, if existing, or a function with the same name, if the variable is not set anymore/yet.

RonaldBarnes replied:
Added `-v` or `-f` to all `unset` commands in #0a86268c4cbf27159536d81e705fabadcb4f5473.

MichaIng commented:
Since aliases are usually added via `~/.bashrc` or `/etc/bash.bashrc`, and hence the script runs on every interactive bash session, isn't it a bit verbose and also huge for the trivial task of a single alias that could be a single line:
```sh
alias occ='sudo -u www-data php /var/www/nextcloud/occ'
```
What I would suggest is using this script as one time alias setup script, which adds the resulting single alias line to `~/.bashrc`. This also renders the whole trap stuff and variable cleanup obsolete, which currently make half of the script.

RonaldBarnes committed:
Specify type of entity being 'unset'.
Run 'unset bash_aliases' to remove the function from environment.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Changed ownership test to file config/config.php, not ./occ

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Added colour to filename when ownership mis-match.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
If no ~/.bash_aliases, then offer to create /etc/profile.d/occ instead.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Offer to add alias to /etc/bash.bashrc as well as ~/.bash_aliases

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes commented:
> Since aliases are usually added via `~/.bashrc` or `/etc/bash.bashrc`, and hence the script runs on every interactive bash session, isn't it a bit verbose and also huge for the trivial task of a single alias that could be a single line like:
> 
> ```shell
> alias occ='sudo -u www-data php /var/www/nextcloud/occ'
> ```

The `bash-add-aliases.sh` is a one-time script.  Assuming the user chooses to have the alias it generates added to some file for persistence.


I like the `~/.bash_aliases` location **much** better than inside `~/.bashrc` as the alias file is often sourced that way (if it exists), and I have a lot of aliases I like to keep separate.

Had to check a bunch of default installs and was surprised this wasn't the standard setup.  It's been so long since I created `~/.bash_aliases` that I don't recall doing it.  And, it's auto-pushed to all my systems, so I expected that to be normal.



> What I suggest is using this script as one time setup script, which adds the resulting single alias line to `~/.bashrc` (or optionally any other defined file). This also renders the whole trap stuff and variable cleanup obsolete (since the script doesn't need to be sourced, but can be executed), which currently make half of the script.

How about, it is a one-time setup script which builds the single alias line, puts it in `~/.bash_aliases` files, if they exist, else puts it into `/etc/bash.bashrc` if user chooses, and keeps the traps so if a user interrupts the procedure, the environment still gets cleaned up?



> But otherwise, especially the bash completion itself: Great stuff +1rocket!!

Thank you!  Really appreciate that.

MichaIng commented:
> I like the `~/.bash_aliases` location much better than inside `~/.bashrc` as the alias file is often sourced that way (if it exists), and I have a lot of aliases I like to keep separate.

Ah you are right, that is even better.

> How about, it is a one-time setup script which builds the single alias line, puts it in `~/.bash_aliases` files, if they exist, else puts it into `/etc/bash.bashrc` if user chooses, and keeps the traps so if a user interrupts the procedure, the environment still gets cleaned up?

`~/.bash_aliases` could be also created, if it's missing. Writing to `/etc/bash.bashrc` of course requires root permissions, but a nice optional feature, to enable the alias for all users.

Btw, `bash-completion` also reads from `~/.local/share/bash-completion/completions`, so the completion itself can be also installed on user-level, if the script is not executed as root, or it is not wanted for all users.

So overall, also to allow running the script in shared hostings, it would be great if the script used the user-level directories for both, alias and completion, and the system-wide dirs only when executed as root (and probably still optionally).
______
Ah right, it does things already as one-time-script. Then I wonder whether it's worth it being sourced and setting up everything for the current shell at all? Is it worth the hassle (and still not ruled out 100% risk that the current shell is unintentionally manipulated by the variables, functions and traps) to avoid the need for a `exec bash` after it finished?

RonaldBarnes commented:
> `~/.bash_aliases` could be also created, if it's missing. Writing to `/etc/bash.bashrc` of course requires root permissions, but a nice optional feature, to enable the alias for all users.

I thought about creating the file, but I checked a Debian Bullseye (11) and it does not source an aliases file. Same with Fedora 37. Ubuntu 22.04 both defines aliases in `~/.bashrc` and sources `~/.bash_aliases` if it exists; both the best and worst choices.


> Btw, `bash-completion` also reads from `~/.local/share/bash-completion/completions`, so the completion itself can be also installed on user-level, if the script is not executed as root, or it is not wanted for all users.

That's good to know, I was not aware.


> So overall, also to allow running the script in shared hostings, it would be great if the script used the user-level directories for both, alias and completion, and the system-wide dirs only when executed as root (and probably still optionally).

Initially, I tried to keep changes mostly in the invoking user's environment.

I will consider offering to put `complete.occ` into `~/.local/share/bash-completion/completions/` and how best to give the user the option between there and the global `/etc/bash_completion.d`.

However, since running `occ` requires the user to have some `sudo` privileges, I don't think having root access as a requirement is critical? Or, in a shared hosting situation, is `config/config.php` owned by the user, not the web server (or does it have 0664 permissions))?

Maybe if user has sudoers privileges to run some things as www-data but writing to `/etc/bash_completion.d/` fails, then I could offer to write to `~/.local/share/bash-completion/completions`?




> Ah right, it does things already as one-time-script. Then I wonder whether it's worth it being sourced and setting up everything for the current shell at all?

I think so.  When gaining shell access first time, run the script, and alias and completions are ready to help with the task at hand.


> Is it worth the hassle (and still not ruled out 100% risk that the current shell is unintentionally manipulated by the variables, functions and traps) to avoid the need for a `exec bash` (or logout/login) after it finished?

I'm not clear on this part.

I may be overlooking something, but the only bit of remaining code that I have concern about is the global var `COMP_WORDBREAKS` having the colon stripped out.

That may affect something somehow somewhere, but I haven't encountered any other completion that depends on it being there and I have not been able to get the completions to work if I `unset` the changes I made to that variable.

(The change affects how the `complete` built-in works, which exists outside the `_occ` function entirely, unfortunately.  Initially expected it worked with the `compgen` built-in, but no such luck.

RonaldBarnes commented:
Interesting issue: if `complete.occ` is copied to `/etc/bash_completion.d/`, it works.
If it's copied to `~/.local/share/bash-completion/completions/`, it does **not** work.  In that location, it's required to be named one of the following:
* occ
* occ.bash
* _occ

I now copy** it to `~/.local/share/bash-completion/completions/occ` but I really don't like the idea of having a different file name, in case someone goes looking for `complete.occ` on their system.

** Only copy it to `~/.local/share/bash-completion/completions/` if it fails to copy to `/etc/bash_completion.d/occ`, which is the most likely target I think, or `/etc/bash_completion.d/` does not exist.


I'm wondering if it's best to just rename `complete.occ` to `occ.bash` now, so wherever it or its copies reside, they always have the same name?

RonaldBarnes commented:
> Ah I see, `~/.bash_aliases` is not sourced by bash directly, but only indirectly via `~/.bashrc` if distro default or admin added it. Then I wouldn't bother with `~/.bash_aliases` at all, as it's basically not guaranteed to work, but add it only to `~/.bashrc`, which is always loaded by bash directly.

The work has been done; I really don't want to undo it.

Plus, `~/.bash_aliases` is the better location for aliases even if it's inconsistently used.
Similar to `~/.bash_profile`, `~/.bash_logout`, and `/.bash_history`.

I should add an option to add the alias to the end of `~/.bashrc` if `~/.bash_aliases` is not found though.


> > I don't think having root access as a requirement is critical?
> 
> It really depends on the environment. In shared hostings, the provided login user may be the webserver user at the same time, in which case sudo is not required at all. In other cases, the login user might have sudo permissions, but probably not for root. But maybe too many checks aren't great either, at least not for now. We can react to specific environments once users face issues with the script in their particular environment, and otherwise assume/require a login user with sudo permissions, or root itself, for now.

I haven't had to deal with shared hosting for a long time, but I agree with your conclusion. 

There'll be a lot of permutations of deployed environments that will reveal issues to work out when they appear.



> Choice for whether to add globally or for current user resp. SUDO_USER only probably via command-line option and/or interactive `read` (as done for other inputs already)?
> 
> > I'm not clear on this part.
> 
> I mean, large parts of the scripts deal with traps and unsetting variables, to not mess with the current shell (the script is sourced into) which is all not required when the script is executed instead of sourced. The only downside is that one cannot run `occ` immediately after executing the script, but needs to refresh the shell session to have the alias and bash completion loaded. This however is pretty common. A hint that one needs to logout/login/refresh/open a new shell session, easiest via `exec bash` to replace the current bash session with a new one.

I wouldn't have thought that an issue. User logs in to server, runs bash script, issues `occ` commands, logs out.  Don't want to force them to re-start bash; I just don't see any benefit but a lot of friction.


> But I do not think there are cases where one wants to load the alias + completion only once, and not install it for future logins.

That seems possible, and is supported.

RonaldBarnes committed:
Option to add completion script to ~/.local/share/bash-completion/completions.
When copying complete.occ, target file is called 'occ'.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
If no ~/.bash_aliases, offer to append to ~/.bashrc instead of /etc/bash.bashrc

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Updated documentation / usage.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Added more web server user names, per https://docs.nextcloud.com/server/22/admin_manual/configuration_server/occ_command.html#http-user-label

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Added @author, reformatted @copyright.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes commented:
Thinking about shared hosting scenarios, maybe instead of throwing error if `config/config.php` is not owned by the web server user, give a warning message then create the alias with the `sudo --user=` having a value of the owner of the `config/config.php` file?

RonaldBarnes committed:
Refactored: removed all global persistence options, stores alias to ~/.bash_aliases or ~/.bashrc and completion file to ~/.local/share/bash-completion/completions

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Changed completion script destination directory to /etc/bash_completion.d/
Renamed completion script occ.bash

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Renamed complete.occ to occ.bash

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes commented:
> Btw, `bash-completion` also reads from `~/.local/share/bash-completion/completions`, so the completion itself can be also installed on user-level, if the script is not executed as root, or it is not wanted for all users.

This does not seem to work consistently, if at all.

Tried in Ubuntu, Debian, and Fedora containers, using the appropriate naming convention (`_occ`, `occ`, or `occ.bash`).

Only got consistent results when copied to `/etc/bash_completion.d/`, although I experienced issues with Fedora and further feedback welcome.

MichaIng commented:
Which OS and bash-completion version did you test with? It seems to be not consistent: https://serverfault.com/a/831184/577419

So until v2.8 (e.g. Debian Buster), `~/.bash_completion.d` seems to work, while on newer version, `~/.local/share/bash-completion/completions` seems to work. But see other answers in the Server Fault question, where the latter should also work (and is recommended) with older version. Probably the official README was a bit behind of what really works. I'll also do some tests here.

MichaIng commented:
It works pretty well here on Debian Buster, Bullseye and Bookworm:
```console
root@VM-Buster:~/.local/share# head -9 bash-completion/completions/7za
# 7z(1) completion                                         -*- shell-script -*-

_7z()
{
    local cur prev words cword
    _init_completion -n = || return

    if [[ $cword -eq 1 ]]; then
        COMPREPLY=( $( compgen -W 'a b d e l t u x test' -- "$cur" ) )
root@VM-Buster:~/.local/share# 7za
a     b     d     e     l     t     test  u     x
```
Also if the file is named `7za.bash`, it works.

RonaldBarnes commented:
> It works pretty well here

I think I found source of one issue I was having:

> PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"

```
root@48dbdfb5bf3a:~# complete -p occ
-bash: complete: occ: no completion specification
root@48dbdfb5bf3a:~# occ [tab][tab]
complete -p occ
complete -F _occ occ
```
So, it's "lazy loading" the completion.

Something to do with 
```
# set up dynamic completion loading
_completion_loader()
  ...
  __load_completion()
```
in `/usr/share/bash-completion/bash_completion`, which is entirely inscrutable.

Interestingly, I'd added an `echo` statement to some versions of `occ.bash` to indicate which location was being invoked, and the "lazy loading" does not display the output.

Sourcing the file does, however, display `echo` output. May have tricked me into thinking it hadn't run.

RonaldBarnes committed:
Typo in usage section.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Restored chown and chmod on /etc/bash_completion.d/occ.bash

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

kesselb commented on bash-add-alias.sh:
```diff
@@ -0,0 +1,298 @@
+		if [[ ${answer} =~ ^Y|y ]] ; then
+			echo "Y"
+			echo ""																	>> ${home_dir}/${alias_file}
+			echo "## tab completion for NextCloud:"	>> ${home_dir}/${alias_file}
```
```suggestion
			echo "## tab completion for Nextcloud:"	>> ${home_dir}/${alias_file}
```

RonaldBarnes replied:
All NextCloud instances changed to Nextcloud.

kesselb commented on bash-add-alias.sh:
```diff
@@ -0,0 +1,298 @@
+
+function _occ_get_nc_path()
+	{
+	read -ep "Path to NextCloud directory? " -i "/" _occ_nc_path
```
```suggestion
	read -ep "Path to Nextcloud directory? " -i "/" _occ_nc_path
```

kesselb commented on bash-add-alias.sh:
```diff
@@ -0,0 +1,298 @@
+	unset -v user_name
+	unset -v home_dir
+	unset -v _occ_alias_exists
+	unset -v phpFound
```
Is there a reason to use camel case for phpFound but snake case for everything else?

RonaldBarnes replied:
I prefer camel case, bash seems to use snake case, and I kept switching projects ending up with mixture.

`phpFound` was the one camel case var name I hadn't changed.

Fixed it now.

RonaldBarnes committed:
NextCloud to Nextcloud, camel case to snake case: phpFound to php_found

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

kesselb commented:
Thanks for your pull request :+1: 

It looks like a nice feature to me.

If I understand the code correct, it's not necessary that the script itself is part of nextcloud server. I think such a helper script should live in an own repository.

RonaldBarnes committed:
Switched back to ~/.local/share/bash-completion/completions/ as the target for occ.bash completion script.
This will 'lazy load' the file once user first tries to use it.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Cleaned up TRAPs; works better with 'read' aka 'readline'.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Added support for 'occ config:...': handles 'app' and 'system', 'get' and 'set'.
If an accepted argument is 'file', handles that as well as 'app', 'user_id', and 'lang'.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes commented:
Just installed "preview generator" in a web browser.

Switched to an existing, open terminal on server, and the auto-competion for it was immediately available and working.

Including `occ preview:generate-all [tab][tab]` giving me a list of site's users.


That's actually pretty nice!

BobJonkman commented:
> If I understand the code correct, it's not necessary that the script itself is part of nextcloud server. I think such a helper script should live in an own repository.

The auto-completion helper script *should* be part of Nextcloud server.

I manage several small Nextcloud instances. I rarely need to use command-line occ, but when I do, auto-completion would be invaluable (because I haven't memorized the sub-commands since I use occ so infrequently).  But because I use occ infrequently, I'm unlikely to hunt down and install an auto-completion helper script separately. So having it as part of the Nextcloud server install  is the best solution.

--Bob.

MichaIng commented on occ.bash:
```diff
@@ -0,0 +1,410 @@
+	unset occ_apps_list
+	}
+
+complete -F _occ occ
```
Is there a way to complete the command based on completion file name or so? On DietPi, we traditionally use `ncc` as Nextcloud `occ` command while `occ` is used for ownCloud's `occ` command, so that (relevant for testing only, of course) both can co-exist. To cover this, generally to allow using any other alias for Nextcloud's occ command, something like `complete -F _occ $0` would be nice, which of course also means that the internal `occ` calls would need to be replaced.

RonaldBarnes replied:
I like using `ncc` instead of `occ`, however:

`apt show ncc`:

> Description: C source code analyzer
> A small, extensible tool that provides program analysis data of C
> source code. This data can be used to analyze the function call flow
> of programs with tools such as graphviz and springgraph. Included
> with this distribution is nccnav that can be used to study the call
> hierarchy of C source code on a text-mode console.

If Nextcloud changes `occ` to `ncc`, I'll take a chance on conflicting completions for `ncc`.

I **can** do a PR to DietPi with the renaming to `ncc` - that's a cool project.

I don't think Nextcloud would accept a PR to to rename `occ` to `ncc`, even if `occ` was then symlinked back to `ncc` for transitional period?



> To cover this, generally to allow using any other alias for Nextcloud's occ command, something like `complete -F _occ $0`

That won't work.  First of all, I believe it tries to use `_occ` function as completion for `complete` itself.

Second of all, running it gives:

```
complete -F _occ $0
-bash: complete: -h: invalid option
complete: usage: complete [-abcdefgjksuv] [-pr] [-DEI] [-o option] [-A action] [-G globpat] [-W wordlist]  [-F function] [-C command] [-X filterpat] [-P prefix] [-S suffix] [name ...]
```

MichaIng replied:
Oh, okay, that is a good argument to stop using `ncc`.

> complete -F _occ $0

I don't know how the bash completion scripts are invoked, so this was just an example to show what I mean: The last argument is the name of the command to complete, which is currently hardcoded to `occ`. Just in case the invoking command matches the name of the script, `$0` would work, but more likely it needs to be `${0##*/}` or doesn't work at all. I don't know whether it can be achieved easily, but if there is a way to have the same completion script work for any alias, just based on the name of the script, that would be awesome, giving users some flexibility about the name of the alias.

However, it's an edge case I guess, so if no one knows this, don't waste time on it. I can investigate myself and open another PR when I find a solution.

RonaldBarnes replied:
> that is a good argument to stop using `ncc`.

I do think `occ` should be `ncc` - it's a good idea, and I wonder how widespread the C source code analyzer `ncc` is?

But the risk of conflict is there and will probably bite someone.


> I don't know whether it can be achieved easily, but if there is a way to have the same completion
> script work for any alias, just based on the name of the script, that would be awesome, giving
> users some flexibility about the name of the alias.

I don't think that trick would work with the `complete` built-in, as I understand things.

The `occ.bash` expects a php file named occ, and the bash script expects a completion script called occ.bash, but one could run `sed -i 's/occ/ncc/g'` on them along with a `mv occ.bash ncc.bash; mv occ ncc`.


I didn't find a repo for DietPi's Nextcloud; if there is one, I'd make the changes required to make it work there.

MichaIng replied:
The PHP file name of course needs to stay the same, else one faces integrity check errors. The alias, which the completion requires, can be freely chosen (not using the bash script but an own bashrc one-liner). The `sed` would replace too much, and I was hoping for a way which does not involve editing this file. Also, if it's shipped with NC core, the issue with integrity check applies for it as well. However, as its not trivial or not possible, for our use case we'll find a solution 🙂. Great to have a bash-completion for occ in any case 👍.

RonaldBarnes replied:
> Is there a way to complete the command based on completion file name or so? On DietPi, we traditionally use `ncc` as Nextcloud `occ` command while `occ` is used for ownCloud's `occ` command

Hi @MichaIng

I think I now understand this issue better.

I run two NC instances on one host.  I created two aliases (one for each full path to a NC install), but `occ.bash` needed tweaks to work properly with being invoked from different aliases.

I have addressed that issue with https://github.com/nextcloud/server/pull/35451/commits/ad15e5e7736119d6fe13684f88d564f39d958d0c.

The creation of 2 aliases is up to the users (see instructions at end of `occ.bash`), as well as adding `complete -F _occ occ_alt_alias` (see instructions at end of `occ.bash`).

MichaIng replied:
Fantastic, many thanks for adding this! It doesn't solve the issue if `occ` itself is used otherwise already, when adding additional `complete` call, one can also call `complete -r occ` in this case.

MichaIng commented on occ.bash:
```diff
@@ -0,0 +1,410 @@
+			fi
+		done
+		## Don't add CWORD, it can be just "-":
+		if [[ ${foundOpt} == "no" && ${value} != ${CWORD} ]] ; then
```
As shell perfectionist 😅, I suggest to consequently use single quotes where no variable expansion is needed. Single quoted strings are taken literally, double quoted strings are going through a bunch of expansion rules in shells. So it is a tiny performance aspect as well.

And the by times multiple empty lines look somehow unnecessary and arbitrary. Probably 1 or 2, consistently, around functions is sufficient.

I won't be able to actually test it until January W2, don't wait for me as it's a little close before next Nextcloud releases 24.0.9 and 25.0.3 and 2023-01-12. Probably we can get it merged for those already 🙂.

RonaldBarnes replied:
When I was writing a lot of Perl, I was really strict about double vs single quotes. Not so much any more.

I'll review the situation with single vs double quotes, but it's really cosmetic, several instances require double quotes for interpolation, and where there are double quotes that can be switched to single quotes, there aren't any unused expansion rules being invoked.

i.e. `if [[ ${foundOpt} == "no" ...` uses double quotes but does not invoke expansion.

MichaIng replied:
Yes it's best practise and a marginal performance aspect only. In most cases (all without spaces in string) you do not need the quotes at all, but they prevent unnecessary globbing attempts. Best argument IMO to protect against possible future failures when others touch the script. The more explicit the syntax, the less the chance that something breaks accidentally.

RonaldBarnes replied:
I looked into changing double quotes to single quotes, but I think it would look weird, i.e.:

```
echo 'Y'
eval alias "occ=${_occ_alias_string}"
```

One line would have single quotes (or no quotes), very next line would have double quotes. That just looks odd.


Also, anyone making changes that isn't aware of interpolation / expansion being restricted to double-quoted strings may stick a variable inside a single-quoted string and wonder why it's output un-expanded. Maybe?


I'm kind of unsure about this one.


I do notice an inconsistency here:

```
green="\e[1;32m"
yellow='\e[1;33m'
red='\e[1;31m'
default_colour="\e[00m"
```

Oops.  Leaning to changing them to double quotes.

RonaldBarnes committed:
Option to regenerate alias if existing one found (helpful if multiple Nextcloud instances installed).

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Consistent use of double vs single quotes.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Changed target for occ.bash to user's ~/.local/share/bash-completion/completions from /etc/bash_completion.d/ -- overlooked it earlier.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes commented:
Question for @artonge - have I overlooked any requested changes?

I think they're all resolved.

Documentation PR at https://github.com/nextcloud/documentation/pull/9478.

Unsure whether I need to add hashes to `core/signature.json`, and if so, how to sign the file.


Thanks!

RonaldBarnes committed:
Allow occ.bash to operate from multiple aliases for admins with multiple Nextcloud instances per machine.
Each instance should have its own alias.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Check for uid as well as user_id (dav:list-calendars: uid; files-scan: user_id).

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

MichaIng commented on occ.bash:
```diff
@@ -0,0 +1,426 @@
+	## When fetching, say, app list, it's critical that correct alias invoked
+	## i.e. Alias `occ` calls main NC v24 & alias `occbclug` calls NC v25
+	occ_alias_name=${1}
+#echo -e "_occ() occ_alias_name=\"${occ_alias_name}\" "
```
Left debug code?
```suggestion
```

RonaldBarnes replied:
I saw the email notification of this comment, fixed it (thank you!), pushed the update, then came here to comment and saw there are a couple buttons where I could have accepted your change.

I ought to learn github interface better...

Thanks @MichaIng - good catch and fixed the issue.

MichaIng replied:
All good, I like this most because it shows the suggested change very clearly, not so much for the ability to easily merge it. The automated commit does not always match project standards anyway.

RonaldBarnes committed:
Removed line of debug code.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

MichaIng commented:
Throwing something new into the ring:
- https://symfony.com/blog/new-in-symfony-5-4-console-autocompletion
- https://symfony.com/doc/current/console.html#console-completion
- https://symfony.com/doc/current/console/input.html#adding-argument-option-value-completion

bash completion natively integrated in Symfony. However, the version used by Nextcloud is too old: https://github.com/nextcloud/3rdparty/blob/master/composer.json
Requires bump from v4.4.30 to v5.4!

RonaldBarnes commented:
Thinking about "discoverability", I personally find the Symfony method a bit hidden (hence I wrote this implementation, only learning about the other method when nearly finished).

Also, I think the script `bash-add-alias.sh` should probably be given a better, more descriptive name, i.e. `occ-bash-completion.sh` or `bash-completion-occ.sh` so an admin looking at the Nextcloud docroot folder will be able to tell at a glance the purpose of the file.

MichaIng commented:
> Thinking about "discoverability", I personally find the Symfony method a bit hidden

It's all a question of documentation. Also your script would need to be documented or found by the admin. In theory, your `bash-add-alias.sh` (I agree the other names would be better) could setup the alias as before but instead of creating an own completion, installing the `symfony-console-completion` one.

But first we should test which one works better or faster. The Symfony one was quite slow. But I'm afraid the major problem is that both call `occ` to derive available commands, which always takes a while to execute. However, if yours works equally good or better (tailored for Nextcloud instead of generic), and is not a maintenance burden, we could remove `symfony-console-completion` from Nextcloud instead.

RonaldBarnes committed:
Changed mode to non-executable: must be sourced for alias to persist after script finishes.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Renamed script to more descriptive name and updated internal usage.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes commented:
> > Thinking about "discoverability", I personally find the Symfony method a bit hidden
> 
> It's all a question of documentation. Also your script would need to be documented or found by the admin.

Having the script file appear in the Nextcloud folder definitely helps with discoverability.

The documentation page for `occ` is very thorough, but it's also easy to miss the completion feature in there.

Having tab completion for a script that's installed outside the system package manager is pretty rare (I can't think of any other example), so one would need to know to look for documentation for a feature they're unlikely to expect to exist.

Also, documentation has been added to a PR to [nextcloud-documentation](https://github.com/nextcloud/documentation/pull/9478).

> In theory, your `bash-add-alias.sh` (I agree the other names would be better)

Agreed! New name: `bash-tab-completion-occ.sh`.

> could setup the alias as before but instead of creating an own completion, installing the `symfony-console-completion` one.


I'm biased, but I much prefer the truncated / grouped completion of my version. Having 100+ options is suboptimal (IMHO).



> But first we should test which one works better or faster. The Symfony one was quite slow. But I'm afraid the major problem is that both call `occ` to derive available commands, which always takes a while to execute. However, if yours works equally good or better (tailored for Nextcloud instead of generic), and is not a maintenance burden, we could remove `symfony-console-completion` from Nextcloud instead.

Yeah, both will be slow due to having to invoke `php` at least once per `tab`, but I'd be interested in performance metrics.


Also, I'm not against the two methods coexisting so users can adopt whichever method they prefer.

MichaIng commented:
> Having the script file appear in the Nextcloud folder definitely helps with discoverability.

We could add a readme with meaningful name to the Nextcloud dir, or, as mentioned, a script to add the alias + Sympfony completion. That the documentation/visibility of an existing implementation is bad is no argument for adding a second implementation but to make the existing implementation visible and documented 😄.

> I'm biased, but I much prefer the truncated / grouped completion of my version. Having 100+ options is suboptimal (IMHO).

THAT is an argument.

> Also, I'm not against the two methods coexisting so users can adopt whichever method they prefer.

Also true. I personally prefer having the core light (even when not noticeable) and such optional features (especially when very rarely used, if at all) as an app. Your variant has the benefit that it doesn't add PHP code (which is at least for CLI requests at least partly loaded) but is independent shell scripts only.

RonaldBarnes committed:
Handle 'dav:move-calendar name sourceuid destinationuid'.
Must fetch all user names before completing 'name' - which is a calendar name, not a user name.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes commented:
Is there anything I still need to do with this PR to get it merged?

> Merging is blocked
> Merging can be performed automatically once the requested changes are addressed.

I *think* I've addressed the requested changes.  If I've missed anything, please let me know.


Thank you

kesselb commented:
> Is there anything I still need to do with this PR to get it merged?

Thank you so much for your contribution. 
We are discussing your pull request. 

In the meantime please

1) Add the newly added files to `build/files-checker.php`
2) Cleanup your git history and reduce the number of commits
3) Rebase the branch to have a recent CI run

RonaldBarnes committed:
Added two script files to build/files-checker.php

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes committed:
Changed references to 'bash-add-alias.sh' to 'bash-tab-completion-occ.sh'.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

kesselb commented on build/files-checker.php:
```diff
@@ -60,6 +60,7 @@
 	'autotest-js.sh',
 	'autotest.sh',
 	'babel.config.js',
+  'bash-tab-completion-occ.sh',
```
Indentation seems off.

RonaldBarnes replied:
Oops, that's embarrassing!

Fixed.

kesselb commented on build/files-checker.php:
```diff
@@ -76,6 +77,7 @@
 	'jest.config.js',
 	'lib',
 	'occ',
+  'occ.bash',
```
Indentation seems off.

RonaldBarnes committed:
Indentation was off; spaces not tabs.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes commented:
> > Is there anything I still need to do with this PR to get it merged?
> 
> Thank you so much for your contribution. We are discussing your pull request.

Thank you!


> In the meantime please
> 
>     1. Add the newly added files to `build/files-checker.php`

Done, and fixed indenting.


>     2. Cleanup your git history and reduce the number of commits
> 
>     3. Rebase the branch to have a recent CI run

Can you give me the `git` commands to do that?  I'm worried about breaking something with `rebase` -- not a `git` expert, by any standard.


Thanks again.

RonaldBarnes committed:
Updated documentation for multi-host, multi-alias environments.

Signed-off-by: Ronald Barnes <ron@ronaldbarnes.ca>

RonaldBarnes commented:
Hello,

I did the requested `git rebase`, I **hope** I didn't make a mess of things!

Also, added minor documentation on multi-instance, multi-alias hosting.

nickvergessen commented:
> I hope I didn't make a mess of things!

I fear you did, but I rebased it again and then force pushed (which was mostlikely what you missed)

RonaldBarnes commented:
> > I hope I didn't make a mess of things!
> 
> I fear you did, but I rebased it again and then force pushed (which was mostlikely what you missed)

Thank you @nickvergessen!

I'd read the docs on `git rebase`, asked a question, waited for a reply, re-read some docs, did the rebase thinking all went well 'til the final step, then...

Appreciate the help & the tip.

nickvergessen commented:
The rebase was mostlikely fine, but afterwards you need to:
```
git push --force origin <branch name>
```

sorbaugh commented:
Hello @RonaldBarnes , fyi I'm reopening this and would like to re-discuss this as the feature looks really nice and looks quite complete.

RonaldBarnes commented:
> fyi I'm reopening this and would like to re-discuss this as the feature looks really nice and looks quite complete.

Thank you, this is great news.

I find this tool extraordinarily handy on several Nextcloud instances running different versions - if someone else in the Nextcloud community can make use of it too, that'd be great!

nickvergessen commented:
- I've been made aware of this again in https://github.com/nextcloud/server/issues/63793
- As pointed out there we are planning to transition from https://github.com/stecman/symfony-console-completion to the native symfony/console https://symfony.com/doc/current/console.html#console-completion
- That should obsolete the occ.bash part mostly
- The only remaining problem is `sudo -u www-data` which also symfony/console does not handle at the moment:
```
$ sudo -u www-data php /var/www/nextcloud/occ completion
Detected shell "nologin", which is not supported by Symfony shell completion (supported shells: "bash", "fish", "zsh").
```

MichaIng commented:
Does `sudo -u www-data SHELL=/bin/bash php ...` work? Or `sudo -u www-data bash -c 'php ...'`?

I assume it tries to determine the shell to apply the completion to from the current shell (which does not exist, as `php` is called directly via `sudo`), and falls back to the current user's default shell, which is commonly `nologin` for system users. So we just need to find out how Symfony detects the current/target shell, which is most likely the `SHELL` variable, and worst case wrap it all into a shell session.

nickvergessen commented:
> EDIT: https://symfony.com/doc/current/setup/symfony_cli.html#symfony-cli-autocompletion
Seems it takes the target shell as argument as well, if the symfony CLI syntax is the same:

I figured that part as well, but pressing `… occ a<tab>` does not do anything still

MichaIng commented:
Of course completion cannot apply through `sudo` or `php`, but only works on the first command. So an alias `alias occ='sudo -u www-data php /path/to/occ'` is needed in any case, and the bash completion function needs to apply to the name of that alias.
